modules/tinymce/src/themes/silver/main/ts/ui/core/complex/BespokeNumberInput.ts

```diff
@@ -0,0 +1,127 @@
+import { AddEventsBehaviour, AlloyComponent, AlloyEvents, Behaviour, Button, Focusing, Input, Keying, NativeEvents, Representing, SketchSpec } from '@ephox/alloy';
+import { Cell, Fun, Id, Optional } from '@ephox/katamari';
+import { Focus, SugarElement, Traverse } from '@ephox/sugar';
+
+import Editor from 'tinymce/core/api/Editor';
+import { UiFactoryBackstage } from 'tinymce/themes/silver/backstage/Backstage';
+
+import { onControlAttached, onControlDetached } from '../../controls/Controls';
+import { updateMenuText, UpdateMenuTextEvent } from '../../dropdown/CommonDropdown';
+import { onSetupEvent } from '../ControlUtils';
+import { NumberInputSpec } from './FontSizeBespoke';
+
+interface BespokeSelectApi {
+  readonly getComponent: () => AlloyComponent;
+}
+
+const createBespokeNumberInput = (editor: Editor, _backstage: UiFactoryBackstage, spec: NumberInputSpec): SketchSpec => {
+  const currentValue = Cell('');
+  const currentComp: Cell<Optional<AlloyComponent>> = Cell(Optional.none());
+
+  const onSetup = onSetupEvent(editor, 'NodeChange', (api: BespokeSelectApi) => {
+    const comp = api.getComponent();
+    currentComp.set(Optional.some(comp));
+    spec.updateText(comp);
+  });
+  const getApi = (comp: AlloyComponent): BespokeSelectApi => ({ getComponent: Fun.constant(comp) });
+  const editorOffCell = Cell(Fun.noop);
+
+  const customEvents = Id.generate('custom-number-input-events');
+
+  const isValidValue = (value: number): boolean => value >= 0;
+
+  const changeValue = (f: (v: number, step: number) => number): void => {
+    const text = currentValue.get();
+    const value = parseFloat(text.match(/^[\d\.]+/)?.join('') ?? '0');
+    const unitRegexp = new RegExp(`(?<=${value})\\D+$`);
+    const unit = text.match(unitRegexp)?.join('') ?? '';
+    const newValue = f(value, spec.getConfigFromUnit(unit).step);
+    const newValueWithUnit = `${isValidValue(newValue) ? newValue : value}${unit}`;
+
+    spec.onAction(newValueWithUnit);
+    currentValue.set(newValueWithUnit);
+    currentComp.get().each((comp) => Representing.setValue(comp, newValueWithUnit));
+  };
+
+  const buttonStyles = {
+    'width': '20px',
+    'text-align': 'center',
+    'background-color': 'grey'
+  };
+
+  return {
+    uid: Id.generate('number-input-wrapper'),
+    dom: {
+      tag: 'div',
+      styles: {
+        display: 'flex'
+      },
+      classes: [ 'tox-number-input' ]
+    },
+    components: [
+      Button.sketch({
+        dom: {
+          tag: 'button',
+          styles: buttonStyles,
+          innerHtml: '-',
+        },
+        action: () => changeValue((n, s) => n - s)
+      }),
+      Input.sketch({
```

Comment:
Based on the other comment, when you make this a memento, you'll pull this out of this object and just have a variable here. You might also want to have a reusable function for the buttons. Then the DOM structure of this becomes much easier to see: 

```
components: [
  makeStepperButton('-', (n, s) => n - s),
  memInput.asSpec(),
  makeStepperButton('+', (n, s) => n + s)
]
```

Reply:
done